gcloud/logging/handlers/transports/background_thread.py

```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Transport for Python logging handler that uses a background worker to log
+to Stackdriver Logging asynchronously."""
+
+import atexit
+import copy
+import threading
+
+from gcloud.logging import Client
+from gcloud.logging.handlers.transports.base import Transport
+
+
+class _Worker(object):
+    """ A threaded worker that writes batches of log entires to the logger
+    API.
+
+    This class reuses a single :class:`Batch` method to write successive
+    entries.
+
+    Currently, the only public methods are constructing it (which also starts
+    it) and enqueuing Logger (record, message) pairs.
+    """
+
+    def __init__(self, logger):
+        self.started = False
+        self.stopping = False
+        self.stopped = False
+
+        # _entries_condition is used to signal from the main thread whether
+        # there are any waiting queued logger entries to be written
+        self._entries_condition = threading.Condition()
+
+        # _stop_condition is used to signal from the worker thread to the
+        # main thread that it's finished its last entries
+        self._stop_condition = threading.Condition()
+
+        # This object continually reuses the same :class:`Batch` object to
+        # write multiple entries at the same time.
+        self.logger = logger
+        self.batch = self.logger.batch()
+
+        self._thread = None
+
+        # Number in seconds of  how long to wait for worker to send remaining
+        self._stop_timeout = 5
+
+        self._start()
+
+    def _run(self):
+        """_run is the entry point for the worker thread. It loops
+        until self.stopping is set to true, and commits batch entries
+        written during :meth:`enqueue`"""
+        try:
+            self._entries_condition.acquire()
+            self.started = True
+            while not self.stopping:
+                if len(self.batch.entries) == 0:
+                    # branch coverage of this code extremely flaky
+                    self._entries_condition.wait()  # pragma: NO COVER
```

Comment:
Sure fair concern. So if it gets stuck on the `wait` for some reason the CPU won't spin, it just won't log. Spinning the CPU would happen if the thread was not shutting down but for some reason `len(self.batch.entries)` wasn't being reset to 0 after the commit ,or if `len` returned <0 which I think is impossible. Otherwise it will just get stuck in the wait until a notify which is only called atexit or when a log message is sent.

`del self.entries[:]` is called at the end of `Batch.commit`, so I guess one fear might be is if for some reason that wasn't called. Maybe we should put the del in a try: finally to prevent that but I think the exception would bubble up and kill the entire worker thread anyway. If the network calls just hang then the thread will just hang.

So I believe that your concern won't be an issue but I understand multi-threaded code is the easiest to introduce bugs in so if we can brainstorm any tests that would alleviate your concerns I can try to write them. FWIW there are both system tests for this code and the unit tests start a real thread (don't mock it). Maybe I could also add a system test that simulates the network outage but it's not clear what I would verify in the test...commit only called a certain number of times? Do syscalls to check CPU usage somehow (feel like that would break portability of tests too much). Or I could even just try to do a manual simulation of some of this stuff myself and screencast it. Of course to do non-determinism of threads I can't prove anything but if you have suggestions I will do my best.